File: frontend/src/api/engine.ts
// Server-side engine evaluation — deeper, on-demand analysis.
import { apiClient } from './client'

export interface ServerLine {
  rank: number
  depth: number
  score_cp: number // centipawns, side-to-move POV
  mate: number // moves to mate, side-to-move POV, 0 if none
  pv: string[] // UCI moves
}

export interface ServerEvaluation {
  fen: string
  depth: number
  best_move: string
  lines: ServerLine[]
}

/** Evaluate a position on the server engine (deeper than the in-browser one). */
export async function evaluatePosition(
  fen: string,
  depth = 18,
  multipv = 3,
): Promise<ServerEvaluation> {
  const { data } = await apiClient.post('/engine/evaluate', {
    fen,
    depth,
    multipv,
  })
  return data
}

File: internal/api/engine.go
// Engine handlers — on-demand position evaluation for the analysis board.
package api

import (
	"net/http"

	"github.com/chessgoddess/chessgoddess/internal/engine"
	"github.com/gin-gonic/gin"
	"github.com/notnil/chess"
)

type EngineHandlers struct {
	pool *engine.Pool
}

func NewEngineHandlers(pool *engine.Pool) *EngineHandlers {
	return &EngineHandlers{pool: pool}
}

type evaluateRequest struct {
	FEN     string `json:"fen" binding:"required"`
	Depth   int    `json:"depth"`
	MultiPV int    `json:"multipv"`
}

// Evaluate runs a bounded engine search on a single position. The board's
// continuous eval runs client-side (WASM); this is for deeper on-demand dives.
func (h *EngineHandlers) Evaluate(c *gin.Context) {
	if h.pool == nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{"error": "engine not available"})
		return
	}

	var req evaluateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "fen is required"})
		return
	}
	if _, err := chess.FEN(req.FEN); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid FEN"})
		return
	}

	// Clamp depth and lines to protect server CPU.
	depth := req.Depth
	if depth < 1 || depth > 22 {
		depth = 14
	}
	multipv := req.MultiPV
	if multipv < 1 || multipv > 5 {
		multipv = 3
	}

	eval, err := h.pool.Evaluate(c.Request.Context(), req.FEN, depth, multipv)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "evaluation failed"})
		return
	}

	c.JSON(http.StatusOK, eval)
}
